tests/integration/bases.py

```diff
@@ -0,0 +1,207 @@
+"""
+Base classes for integration tests
+"""
+
+import inspect
+import logging
+import os
+import re
+import subprocess
+from pathlib import Path
+from types import FunctionType
+from typing import Iterable, Union
+
+import pandas as pd
+
+from python2verilog import (
+    Modes,
+    get_actual_raw,
+    get_context,
+    get_expected,
+    namespace_to_file,
+    namespace_to_verilog,
+    verilogify,
+)
+from python2verilog.backend.verilog.config import CodegenConfig
+from python2verilog.simulation import iverilog
+from python2verilog.simulation.display import strip_ready, strip_valid
+
+from .utils import make_tuple
+
+
+class BaseTest:
+    def __init_subclass__(cls) -> None:
+        cls.statistics = []
+        cls.write = []
+
+    def __test(
+        self,
+        funcs: Union[Iterable[FunctionType], FunctionType],
+        test_cases: Iterable[Union[tuple[int, ...], int]],
+        config: CodegenConfig,
+    ):
+        if not isinstance(funcs, list):
+            funcs = [funcs]
+
+        if self.args.first_test:
+            test_cases = [next(iter(test_cases))]
+        test_cases = list(map(make_tuple, test_cases))
```

Comment:
How clear is this code?

Reply:
I think allowing multiple ways to define test_cases is making code unnecessarily complex. If you force test cases and funcs to be defined as an array/iterable of tuples, then there's no need for the Union or this statement.

I'm not too sure what this code segment does.

Reply:
> I think allowing multiple ways to define test_cases is making code unnecessarily complex. If you force test cases and funcs to be defined as an array/iterable of tuples, then there's no need for the Union or this statement.

Doing this would eliminate line 48. However, lines 46-47 would still remain. This is for the `pytest` option to only use the first test case in the test cases (i.e. for a quicker run).

Reply:
I see, I would comment that it's for the `pytest` option for additional clarity.

Reply:
Well, I can't be adding an explanation comment every time I use a `pytest` option.

I hope the following explained well enough.

https://github.com/WorldofKerry/Python2Verilog/blob/0efc93249ec704bbb648ba369c2201a284475d0c/README.md#L81

Reply:
I was unaware of those flags. If people know about those flags, then the comment is clear